### PR TITLE
FEATURE: Remove lodash usage and deprecate it.

### PR DESCRIPTION
**Pull request by eviltrout**
This removes lodash usage. Most uses are not useful but some might be worth keeping around. In that case it should be a model.

### CODE REVIEW

File: app/assets/javascripts/discourse/app/components/badge-card.js [hunk 1]
```diff
@@ -25,7 +26,7 @@ export default Component.extend({
   summary(size) {
     if (size === "large") {
       const longDescription = this.get("badge.long_description");
-      if (!_.isEmpty(longDescription)) {
+      if (!isEmpty(longDescription)) {
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/components/badge-card.js:29]: Careful on that one:

File: app/assets/javascripts/locales/i18n.js [hunk 2]
```diff
@@ -323,7 +323,7 @@ I18n.enableVerboseLocalization = function() {
     if (!current) {
       current = keys[scope] = ++counter;
       var message = "Translation #" + current + ": " + scope;
-      if (!_.isEmpty(value)) {
+      if (value && Object.keys(value).length > 0) {
```
Comment (jjaffeux) [inline app/assets/javascripts/locales/i18n.js:326]: Reading this, it seems like you are aware of the difference between the two isEmpty

File: app/assets/javascripts/discourse/app/models/post-stream.js [hunk 3]
```diff
@@ -145,8 +145,12 @@ export default RestModel.extend({
   **/
   @discourseComputed("posts.[]", "stream.[]")
   previousWindow() {
+    if (!this.posts) {
+      return [];
+    }
+
     // If we can't find the last post loaded, bail
-    const firstPost = _.first(this.posts);
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/models/post-stream.js:149]: the only thing first does is working with null/undefined.

```
_.first(null) // undefined
```

So I guess codepaths replacing _.first should ensure the object is not null/undefined.

**Comment by jjaffeux** [comment]
We might also want to remove these functions from https://github.com/discourse/discourse/blob/master/lib/tasks/javascript.rake#L223

Probably with a deprecation.

**Comment by xfalcox** [comment]
There are some plugin uses like https://github.com/discourse/discourse-patreon/blob/master/assets/javascripts/discourse/controllers/admin-plugins-patreon.js.es6#L14

File: app/assets/javascripts/discourse/app/models/post-stream.js (hunk 3 above)
Comment (eviltrout) [reply app/assets/javascripts/discourse/app/models/post-stream.js:149]: I believe I accounted for this. All tests are passing.

**Comment by eviltrout** [comment]
@xfalcox yes I am aware. The PR was still in draft and I hadn't finished at that point.

It now removes all lodash usage in our codebase and deprecates it for plugins/themes to continue working until we update their code too. I've slated it for removal in 2.7.

**Pull request merged**